File: src/main/scala/mrtjp/fengine/tiles/FEPortlessWireTile.scala
package mrtjp.fengine.tiles

/** A directionally connected wire that ignores ports
  *
  *   - Signal can enter from masked direction from any port
  *   - Signal can exit to masked direction to all ports
  */
trait FEPortlessWireTile extends FETile
{
    /** Mask of allowed directions */
    val connMask:Int

    override def propagationFunc(inDir:Int, inPort:Int):(Int, Int) => Boolean = { (outdir:Int, outPort:Int) =>
        (connMask&1<<inDir) != 0 && (connMask&1<<outdir) != 0
    }
}

File: src/main/java/mrtjp/fengine/tiles/FEPortlessWireTile.java
package mrtjp.fengine.tiles;

import mrtjp.fengine.api.PropagationFunction;

public abstract class FEPortlessWireTile implements FETile {

    /**
     * Mask of allowed directions
     */
    protected abstract int getConnMask();

    @Override
    public PropagationFunction propagationFunc(int inDir, int inPort) {
        return ((outDir, outPort) -> (getConnMask() & 1 << inDir) != 0 && (getConnMask() & 1 << outDir) != 0);
    }
}
